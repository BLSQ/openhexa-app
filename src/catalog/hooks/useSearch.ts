import { gql, useQuery } from "@apollo/client";
import SearchResult from "catalog/features/SearchResult";
import { SearchResult_ResultFragment } from "catalog/features/SearchResult.generated";
import { SearchQueryQuery } from "./useSearch.generated";

export type UseSearchOptions = {
  query?: string;
  datasourceIds?: string[];
  types?: string[];
  page?: number;
  perPage?: number;
};

function useSearch(options: UseSearchOptions) {
  const { query, page, perPage, datasourceIds, types } = options;
  const { data, loading, previousData } = useQuery<SearchQueryQuery>(
    gql`
      query SearchQuery(
        $query: String
        $types: [String!]
        $datasourceIds: [String!]
        $page: Int
        $perPage: Int
      ) {
        search(
          query: $query
          datasourceIds: $datasourceIds
          types: $types
          page: $page
          perPage: $perPage
        ) {
          results {
            ...SearchResult_result
          }
          types {
            value
            label
          }
        }
      }
      ${SearchResult.fragments.result}
    `,
    {
      variables: { query, page, perPage, datasourceIds, types },
    }
  );

  const results = (data || previousData)?.search.results;
  return {
    results: results as SearchResult_ResultFragment[] | undefined,
    types: (data || previousData)?.search.types,
    loading,
  } as const;
}

export default useSearch;
