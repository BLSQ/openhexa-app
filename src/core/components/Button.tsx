import clsx from "clsx";
import { ButtonHTMLAttributes, ReactElement } from "react";

const ButtonVariants = ["primary", "secondary", "white", "outlined", "custom"];
type ButtonVariant = typeof ButtonVariants[number];

const ButtonSizes = ["sm", "md", "lg", "xl", "xxl"];
type ButtonSize = typeof ButtonSizes[number];

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  className?: string;
  leadingIcon?: ReactElement;
}

export const Classes: { [key: string]: string } = {
  base: `relative shadow-sm rounded h-fit
    focus:outline-none focus:ring-2 focus:ring-offset-2
    inline-flex items-center justify-center transition-all
    border border-transparent font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
  `,
  primary:
    "border-transparent text-white bg-blue-600 hover:bg-blue-700 focus:ring-blue-500",
  secondary:
    "border-transparent text-indigo-700 bg-indigo-100 hover:bg-indigo-200 focus:ring-indigo-500",
  white: "text-gray-800 border-gray-300 bg-white hover:bg-gray-50",
  outlined: "text-white hover:bg-white hover:text-gray-800 border-white",

  // Sizes
  sm: "text-xs px-2 py-1",
  md: "text-sm px-3 py-2.5 leading-4",
  lg: "text-sm px-4 py-2.5",
  xl: "text-base px-4 py-2",
  xxl: "text-base px-6 py-3",
};

const Button = (props: ButtonProps) => {
  const {
    variant = "primary",
    size = "md",
    className,
    children,
    leadingIcon,
    disabled,
    ...delegated
  } = props;

  const classes = clsx(
    Classes.base,
    Classes[size],
    variant !== "custom" && Classes[variant],
    className
  );
  return (
    <button className={classes} disabled={disabled} {...delegated}>
      {leadingIcon ? (
        <LeadingIcon size={size}>{leadingIcon}</LeadingIcon>
      ) : null}
      {children}
    </button>
  );
};
Button.Variants = ButtonVariants;
Button.Sizes = ButtonSizes;

const LeadingIcon = (props: { children: ReactElement; size?: ButtonSize }) => {
  return (
    <div
      className={clsx("-ml-1", props.size === ButtonSizes[0] ? "mr-1" : "mr-2")}
      aria-hidden="true"
    >
      {props.children}
    </div>
  );
};

export default Button;
