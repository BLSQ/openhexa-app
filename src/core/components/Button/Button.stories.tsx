import { DocumentArrowDownIcon } from "@heroicons/react/24/outline";
import { Story } from "@ladle/react";

import ButtonComponent, { ButtonProps } from "./Button";

interface ButtonStoryProps extends ButtonProps {
  leading: boolean;
}

export const Button: Story<ButtonStoryProps> = ({
  children,
  variant,
  size,
  leading,
}) => {
  const leadingIcon = leading ? (
    <DocumentArrowDownIcon className="h-4" />
  ) : undefined;
  return (
    <ButtonComponent size={size} variant={variant} leadingIcon={leadingIcon}>
      {children}
    </ButtonComponent>
  );
};

Button.storyName = "default";
Button.args = {
  children: "Hello world",
  leading: false,
};
Button.argTypes = {
  variant: {
    options: ButtonComponent.Variants,
    control: { type: "radio" },
    defaultValue: "primary",
  },
  size: {
    options: ButtonComponent.Sizes,
    control: { type: "select" },
    defaultValue: "md",
  },
};

export const ButtonSnapshot: Story = () => (
  <div className="space-y-6">
    {ButtonComponent.Variants.map((variant) => (
      <div key={variant} className="flex items-center gap-4">
        {ButtonComponent.Sizes.map((size) => (
          <ButtonComponent key={size} variant={variant} size={size}>
            {variant}
          </ButtonComponent>
        ))}
      </div>
    ))}
  </div>
);
ButtonSnapshot.storyName = "Snapshot";
ButtonSnapshot.meta = {
  snapshot: true,
};

const defaults = {
  title: "UI / Button",
};

export default defaults;
