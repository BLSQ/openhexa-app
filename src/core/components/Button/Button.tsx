import clsx from "clsx";
import { ButtonHTMLAttributes, ReactElement } from "react";

const ButtonVariants = [
  "primary",
  "secondary",
  "white",
  "outlined",
  "custom",
] as const;
type ButtonVariant = typeof ButtonVariants[number];

const ButtonSizes = ["sm", "md", "lg", "xl", "xxl"] as const;
type ButtonSize = typeof ButtonSizes[number];

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  className?: string;
  leadingIcon?: ReactElement;
}

export const Classes: { [key: string]: string } = {
  base: `relative shadow-sm rounded h-fit
    focus:outline-none focus:ring-2 focus:ring-offset-2
    inline-flex items-center justify-center transition-all
    border font-medium
    disabled:opacity-50 disabled:cursor-not-allowed
  `,
  primary:
    "border-transparent text-white bg-blue-600 hover:bg-blue-700 focus:ring-blue-500",
  secondary:
    "border-transparent text-indigo-700 bg-indigo-100 hover:bg-indigo-200 focus:ring-indigo-500",
  white: "text-gray-800 border-gray-400 bg-white hover:bg-gray-50",
  outlined: "text-gray-800 border-transparent shadow-none hover:bg-gray-100",

  // Sizes
  sm: "text-xs px-2 py-1",
  md: "text-sm px-3 py-2.5 leading-4",
  lg: "text-sm px-4 py-2.5",
  xl: "text-base px-4 py-2",
  xxl: "text-base px-6 py-3",
};

const Button = (props: ButtonProps) => {
  const {
    variant = "primary",
    size = "md",
    className,
    children,
    leadingIcon,
    disabled,
    ...delegated
  } = props;

  const classes = clsx(
    Classes.base,
    Classes[size],
    variant !== "custom" && Classes[variant],
    className
  );
  return (
    <button className={classes} disabled={disabled} {...delegated}>
      {leadingIcon ? (
        <LeadingIcon size={size}>{leadingIcon}</LeadingIcon>
      ) : null}
      {children}
    </button>
  );
};
Button.Variants = ButtonVariants;
Button.Sizes = ButtonSizes;

const LeadingIcon = (props: { children: ReactElement; size?: ButtonSize }) => {
  return (
    <div
      className={clsx(
        "-ml-1",
        props.size === ButtonSizes[0] ? "mr-1" : "mr-1.5"
      )}
      aria-hidden="true"
    >
      {props.children}
    </div>
  );
};

export default Button;
