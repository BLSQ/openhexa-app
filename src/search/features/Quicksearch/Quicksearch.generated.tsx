import * as Types from '../../../graphql-types';

import { gql } from '@apollo/client';
export type Quicksearch_CollectionFragment = { __typename?: 'Collection', id: string };

export const Quicksearch_CollectionFragmentDoc = gql`
    fragment Quicksearch_collection on Collection {
  id
}
    `;