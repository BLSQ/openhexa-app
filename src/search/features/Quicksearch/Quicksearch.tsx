import { SearchIcon } from "@heroicons/react/outline";
import axios from "axios";
import clsx from "clsx";
import Dialog from "core/components/Dialog";
import Input from "core/components/forms/Input";
import Link from "core/components/Link";
import Spinner from "core/components/Spinner";
import useDebounce from "core/hooks/useDebounce";
import { useTranslation } from "next-i18next";
import { ReactElement, useEffect, useState } from "react";

export type QuickSearchResult = {
  id: string;
  object_id: string;
  app_label: string;
  display_name: string;
  content_type_model: string;
  content_type_name: string;
  datasource_name: string;
  datasource_id: string;
  description: string;
  rank: number;
  url: string;
  symbol: string;
};
type QuicksearchProps = {
  open?: boolean;
  onClose(): void;
  renderActions?: (element: QuickSearchResult) => ReactElement;
};

function useQuicksearch(query: string) {
  const [isLoading, setLoading] = useState(false);
  const [data, setData] = useState<QuickSearchResult[]>([]);

  async function fetchResults(query: string) {
    setLoading(true);
    try {
      const response = await axios.get("/catalog/quick-search/", {
        params: { query },
      });
      setData(response.data.results);
    } catch (err) {
      console.error(err);
    } finally {
      setLoading(false);
    }
  }

  useEffect(() => {
    if (!query && data) {
      setData([]);
    }
    if (query) {
      fetchResults(query);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [query]);

  return [data, { isLoading }] as const;
}

const Quicksearch = (props: QuicksearchProps) => {
  const { open = false, onClose, renderActions } = props;
  const [queryString, setQueryString] = useState("");
  const { t } = useTranslation();
  const debouncedQueryString = useDebounce(queryString, 120);
  const [data, { isLoading }] = useQuicksearch(debouncedQueryString);

  useEffect(() => {
    if (!open) {
      setQueryString("");
    }
  }, [open]);

  return (
    <Dialog open={open} onClose={onClose} maxWidth="max-w-2xl">
      <Dialog.Content>
        <Input
          value={queryString}
          onChange={(e) => setQueryString(e.target.value)}
          trailingIcon={
            isLoading ? <Spinner size="xs" /> : <SearchIcon className="h-4" />
          }
          placeholder={t("Search for elements...")}
        />

        <>
          <span
            className={clsx(
              "text-sm text-gray-500 transition-all",
              !data.length && "invisible"
            )}
          >
            {t("Showing top 10 results")}
          </span>
          <div className="mt-4">
            <div className="-mx-6 h-52 overflow-y-auto">
              {data.length === 0 && (
                <div className="mt-4 w-full text-center text-sm italic text-gray-500">
                  {queryString
                    ? t("No results to display")
                    : t("Type what you are looking for here above")}
                </div>
              )}
              {data.map((result) => (
                <div
                  className="flex items-center gap-3 py-2.5 px-7 text-sm hover:bg-gray-50"
                  key={result.id}
                >
                  {result.symbol && (
                    <div className="w-5">
                      <img
                        className="h-5 w-5 flex-shrink-0"
                        src={result.symbol}
                        alt={result.display_name}
                      />
                    </div>
                  )}
                  <div className="flex-1 truncate">
                    <Link className="font-medium" href={result.url}>
                      {result.display_name}
                    </Link>
                    <span className="ml-2 font-sans text-xs text-gray-300">
                      #{Math.round(result.rank * 100) / 100}
                    </span>
                    <div className="mt-1 text-xs text-gray-500">
                      {result.content_type_name}{" "}
                      {result.datasource_name && `/ ${result.datasource_name}`}
                    </div>
                  </div>
                  {renderActions && renderActions(result)}
                </div>
              ))}
            </div>
          </div>
        </>
      </Dialog.Content>
    </Dialog>
  );
};

export default Quicksearch;
