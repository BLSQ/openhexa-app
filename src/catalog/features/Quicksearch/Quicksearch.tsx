import { ArrowRightIcon } from "@heroicons/react/24/outline";
import useSearch from "catalog/hooks/useSearch";
import Dialog from "core/components/Dialog";
import Link from "core/components/Link";
import useDebounce from "core/hooks/useDebounce";
import { useTranslation } from "next-i18next";
import { useRouter } from "next/router";
import { useEffect, useState } from "react";
import SearchInput from "../SearchInput";
import SearchResult, { SearchResultProps } from "../SearchResult";

type QuicksearchProps = {
  open?: boolean;
  onClose(): void;
} & Pick<SearchResultProps, "renderActions">;

const Quicksearch = (props: QuicksearchProps) => {
  const { open = false, onClose, renderActions } = props;
  const [queryString, setQueryString] = useState("");
  const { t } = useTranslation();
  const router = useRouter();
  const debouncedQueryString = useDebounce(queryString, 120);

  const { results, loading } = useSearch({
    query: debouncedQueryString,
    perPage: 10,
    skip: !open,
  });

  useEffect(() => {
    if (open) {
      onClose();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [router.pathname, open]);

  useEffect(() => {
    if (!open) {
      setQueryString("");
    }
  }, [open]);

  return (
    <Dialog
      open={open}
      onClose={onClose}
      maxWidth="max-w-2xl"
      centered={false}
      padding="py-2"
    >
      <Dialog.Content className="space-y-2 divide-y divide-gray-100">
        <SearchInput
          value={queryString}
          loading={loading}
          className="border-none px-4 outline-none focus:ring-0"
          onChange={(e) => setQueryString(e.target.value)}
        />
        {results?.length === 0 && queryString && (
          <div className="p-4 text-sm text-gray-500">
            {t("No results to display.")}
          </div>
        )}

        {results && results.length !== 0 && (
          <div className="overflow-y-auto">
            {results.map((result) => (
              <SearchResult
                key={result.object.id}
                result={result}
                renderActions={renderActions}
              />
            ))}
          </div>
        )}

        <div className="flex justify-between px-4 pt-3 pb-1 text-xs text-gray-500 transition-all">
          {results && (
            <span>{results.length > 0 && t("Showing top 10 results")}</span>
          )}
          <Link
            href={{
              pathname: "/catalog/search",
              query: { query: queryString },
            }}
            className="flex items-center gap-1"
          >
            {t("Go to advanced search")}
            <ArrowRightIcon className="w-3" />
          </Link>
        </div>
      </Dialog.Content>
    </Dialog>
  );
};

export default Quicksearch;
