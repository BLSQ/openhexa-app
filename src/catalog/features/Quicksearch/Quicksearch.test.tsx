import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import useSearch from "catalog/hooks/useSearch";
import { SearchQueryDocument } from "catalog/hooks/useSearch.generated";
import Button from "core/components/Button";
import mockRouter from "next-router-mock";
import { useState } from "react";
import Quicksearch from "./Quicksearch";

jest.mock("catalog/hooks/useSearch", () => jest.fn(() => ({})));
jest.mock("core/hooks/useDebounce", () => jest.fn((v) => v));

const useSearchMock = useSearch as jest.Mock;

describe("Quicksearch", () => {
  const onClose = jest.fn();
  afterEach(() => {
    onClose.mockClear();
  });

  it("is not displayed by default", async () => {
    const { container } = render(<Quicksearch onClose={onClose} />);
    const dialog = await screen.queryByRole("dialog");
    expect(dialog).not.toBeInTheDocument();
    expect(onClose).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it("is directly shown if open is true", async () => {
    const { container } = render(<Quicksearch onClose={onClose} open />);
    const dialog = await screen.queryByRole("dialog");
    expect(dialog).toBeInTheDocument();
    expect(onClose).not.toHaveBeenCalled();
  });

  it("does not fetch results if the query if empty", async () => {
    const { container } = render(<Quicksearch onClose={onClose} open />);
    const dialog = await screen.queryByRole("dialog");
    expect(dialog).toBeInTheDocument();
    expect(onClose).not.toHaveBeenCalled();
    expect(useSearchMock).toHaveBeenCalledWith({
      query: "",
      perPage: 10,
      skip: true, // Tested line
    });
  });

  it("displays the results", async () => {
    const user = await userEvent.setup({ delay: null });
    useSearchMock.mockImplementation(() => {
      return {
        results: [
          {
            rank: 1,
            object: {
              __typename: "Collection",
              id: "1",
              name: "Collection",
            },
          },
        ],
        types: [],
        loading: false,
      };
    });

    const mocks = [
      {
        request: {
          query: SearchQueryDocument,
          variables: {
            input: {
              query: "Test",
              perPage: 10,
            },
          },
        },
      },
    ];

    render(<Quicksearch onClose={onClose} open />);

    const dialog = await screen.queryByRole("dialog");
    expect(dialog).toBeInTheDocument();
    expect(useSearchMock).toHaveBeenCalledWith({
      query: "",
      perPage: 10,
      skip: true, // Tested line
    });
    const searchInput = screen.getByTestId("search-input");
    await user.type(searchInput, "Test");

    await screen.findByDisplayValue("Test");
    expect(useSearchMock).toHaveBeenLastCalledWith({
      query: "Test",
      skip: false,
      perPage: 10,
    });

    expect(await screen.findByTestId("results")).toBeInTheDocument();
  });

  it("removes the dialog from the dom when closed", async () => {
    function TestQuicksearch() {
      const [isOpen, setOpen] = useState(false);

      return (
        <>
          <Quicksearch onClose={() => setOpen(false)} open={isOpen} />
          <Button onClick={() => setOpen(!isOpen)} data-testid="toggle">
            Toggle
          </Button>
        </>
      );
    }
    const user = await userEvent.setup({ delay: null });

    render(<TestQuicksearch />);

    expect(await screen.queryByRole("dialog")).toBeNull();
    const btn = screen.getByTestId("toggle");

    await user.click(btn);
    expect(await screen.findByRole("dialog")).toBeInTheDocument();

    await user.click(btn);
    expect(await screen.queryByRole("dialog")).toBeNull();
  });

  it("calls onClose when the route changes", async () => {
    mockRouter.setCurrentUrl("/");
    render(<Quicksearch onClose={onClose} open />);

    expect(onClose).not.toHaveBeenCalled();

    waitFor(() => {
      expect(screen.queryByRole("dialog")).toBeInTheDocument();
    });

    mockRouter.setCurrentUrl("/search");

    waitFor(() => {
      expect(onClose).toHaveBeenCalled();
    });
  });

  it("redirects the user to the advanced search and forward the querystring", async () => {
    mockRouter.setCurrentUrl("/");
    const user = await userEvent.setup({ delay: null });
    render(<Quicksearch onClose={onClose} open />);

    expect(screen.queryByRole("dialog")).toBeInTheDocument();
    expect(onClose).not.toHaveBeenCalled();

    await user.click(screen.getByText("Go to advanced search"));
    expect(onClose).toHaveBeenCalled();
  });
});
