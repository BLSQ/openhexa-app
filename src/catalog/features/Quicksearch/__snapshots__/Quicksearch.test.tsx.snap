// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quicksearch is not displayed by default 1`] = `<div />`;
